### DISCUSSION TITLE
Query assistance with RepeatUntil

### DISCUSSION THREAD

**Discussion by AnthonyRyck** [Gremlinq/ExRam.Gremlinq#477] [Q&A] [answered]
Hello,  
I would like to make a request with Gremlinq but I am a little lost with RepeatUntil.  
This is my Gremlin query 
```
g.V().has('name', 'startStation')
.repeat(out ().simplePath())
.until(has('name', 'StopStation'))
.path().limit(1)
```
Its to find the best path between 2 stations (Vertex).  

```
var startStation = await _g.V<StationVertex>().Where(x => x.StationName == "nameStart")
								.FirstAsync();
var test = await _g.V(startStation.Id).RepeatUntil(__ => __.Out().SimplePath(),
						__ => __.V<StationVertex>().Where(x => x.StationName == "nameStop").Path().Limit(1));
```
The result give me just the 2 vertices StationVertex Out on start station.  
Thanks for your help.

**Comment by danielcweber** [comment]
Call Debug() on the query, it'll return a gremlin string. From there you'll find where the difference to your original query is.

**Reply by AnthonyRyck** [reply]
Thanks for the command :)
And the result :
```
{
	"Script":
	"V(1092).repeat(__.out().simplePath())
			.until(__.V().hasLabel('StationVertex').has('StationName','Reisen').path().limit(1))
			.project('id','label','properties')
			.by(id)
			.by(label)
			.by(__.properties()
					.group()
					.by(label)
					.by(__.project('id','label','value','properties')
							.by(id)
							.by(label)
							.by(value)
							.by(__.valueMap())
							.fold()))",
	"Bindings":{}
}
```

**Reply by AnthonyRyck** [reply]
OK !  
This request :  
```
var temtt = await _g.V(startStation.Id)
                .RepeatUntil(__ => __.Out().SimplePath(),
			__ => __.V<StationVertex>().Where(x => x.StationName == "nameStop"))
                .Path().Limit(1);
```
the result :   
```
V(1092).repeat(__.out().simplePath())
	.until(__.V().hasLabel('StationVertex').has('StationName','nameStop'))
	.path().limit(1)
```
and what I want :  
```
g.V(1092).repeat(out().simplePath())
.until(has('name', 'StopStation'))
.path().limit(1)
```

What is "___" ? Can you give me a "definition"  ?  
*sorry for my english*

**Comment by danielcweber** [comment]
__ is the anonymous traversal. It's also a concept in Gremlin (see https://tinkerpop.apache.org/docs/current/reference/). It starts off where the outer traversal left off. As such, your use of __ => __.V&lt;StationVertex>() is in error, it's not in your original query. Just leave it out but use V&lt;StationVertex> at the top of your query instead.

**Reply by AnthonyRyck** [reply]
Sorry I don't understand when you say: "_Just leave it out but use V<StationVertex> at the top of your query instead._"  
Thanks you for your help.  
The project is great. I think that when I master the writing, it will be a killing ;)

**Comment by danielcweber** [comment, marked as answer]
The `V<StationVertex>` mid-query is not in the original query:

![image](https://user-images.githubusercontent.com/1899410/136158885-938fe632-341b-453b-8a0f-e34f6d5da13a.png)

That's why the resulting Gremlin-query differs in exactly that spot!

Try 

```
var temtt = await _g
                .V<StationVertex>(startStation.Id)
                .RepeatUntil(__ => __.Out().SimplePath(),
			__ => __.Where(x => x.StationName == "nameStop"))
                .Path().Limit(1);
```

instead which should be more what you intended and see what it does.
